### ISSUE TITLE
SQLFluff breaks when using DBT macros: jinja2.exceptions.UndefinedError: 'dbt_utils' is undefined

### ISSUE THREAD

**Issue by megasend** [sqlfluff/sqlfluff#4207]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Hi there!
First, I want to thank you for a great tool!

I'm trying to use SQLFluff in a DBT project, but it looks like SQLFluff is breaking when trying to parse the `dbt_utils` macros:
```
select
  {{ dbt_utils.surrogate_key(['order_id']) }} as uniq_id,
  -- etc.
```

The rest of the DBT-specific Ninja syntax is recognized without problems. If remove the use of the `dbt_utils` macros, then the SQLFluff also works without problems. I don’t know if this is important or not, but if to comment the above line in the SQL query, the error is the same.

### Expected Behaviour

No errors expected

### Observed Behaviour

I configured SQLFluff according to the documentation: installed the dbt plugin and added the appropriate configs, which are listed below.

### How to reproduce

DBT's packages.yml:
```
packages:
  - package: dbt-labs/dbt_utils
    version: 0.8.5
```

DBT model (extra details removed for simplicity):
```
{{
  config(
    sort='date_created',
    dist='customer_id'
  )
}}

select
  {{ dbt_utils.surrogate_key(['order_id'])}} as uniq_id,
  customer_id,
  date_created
from {{ref('shopify_order')}}
```

cmd:
`sqlfluff lint --verbose --config .sqlfluff ./models/shopify_order.sql`

Stacktrace:
```
==== sqlfluff ====
sqlfluff:                1.4.1 python:                 3.8.10
implementation:        cpython verbosity:                   1
dialect:             snowflake templater:                 dbt
dbt:                    =1.2.0

==== readout ====

=== [ path: ./models/shopify_order.sql ] ===

effective configured processes:  4
=== [dbt templater] Sorting Nodes...
 
=== [dbt templater] Compiling dbt project...
=== [dbt templater] Project Compiled.
ERROR      Fatal dbt compilation error on /home/mark/projects/warehouse/models/shopify_order.sql. This occurs most often during incorrect sorting of ephemeral models before linting. Please report this error on github at https://github.com/sqlfluff/sqlfluff/issues, including both the raw and compiled sql for the model affected. 
Traceback (most recent call last):
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 494, in catch_jinja
    yield
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 526, in render_template
    return template.render(ctx)
  File "/home/mark/.local/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/home/mark/.local/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/home/mark/.local/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 18, in top-level template code
  File "/home/mark/.local/lib/python3.8/site-packages/jinja2/sandbox.py", line 407, in getattr
    value = getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'dbt_utils' is undefined

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/mark/.local/lib/python3.8/site-packages/sqlfluff_templater_dbt/templater.py", line 481, in _unsafe_process
    node = self.dbt_compiler.compile_node(
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/compilation.py", line 515, in compile_node
    node = self._compile_node(node, manifest, extra_context)
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/compilation.py", line 374, in _compile_node
    compiled_node.compiled_sql = jinja.get_rendered(
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 571, in get_rendered
    return render_template(template, ctx, node)
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 526, in render_template
    return template.render(ctx)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/mark/.local/lib/python3.8/site-packages/dbt/clients/jinja.py", line 499, in catch_jinja
    raise UndefinedMacroException(str(e), node) from e
dbt.exceptions.UndefinedMacroException: Compilation Error in model shopify_order (models/shopify_order.sql)
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".
WARNING    Skipped file /home/mark/projects/warehouse/models/shopify_order.sql because dbt raised a fatal exception during compilation: Compilation Error in model shopify_order (models/shopify_order.sql)
  'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps". 
```

### Dialect

snowflake

### Version

```
sqlfluff --version
sqlfluff, version 1.4.1
```

```
python3 --version
Python 3.8.10
```

```
pip show sqlfluff-templater-dbt
Version: 1.4.1
```

```
pip show dbt-core
Version: 1.2.0
```

### Configuration

**.sqlfluff:**
```
[sqlfluff]
ignore = templating
templater = dbt
dialect = snowflake
processes = 4
ignore_templated_areas = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:templater:dbt]
profile = default
target = prod

[sqlfluff:rules]
tab_space_size = 2
max_line_length = 110
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all
```

**.sqlfluffignore:**
```
target/
dbt_modules/
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by moreaupascal56** [comment]
Hello just trying to help as I can, I tried your config on my side and had no issues. 
The error looks like it is a dbt compilation error more than a sqlfluff error, is `dbt compile` working ? Could as well be because the package dbt_utils is not installed correctly, so a `dbt deps` could maybe help?

**Comment by moreaupascal56** [comment]
Hello I had a further look and all I was able to found is this [slack discussion](https://getdbt.slack.com/archives/C2JRRQDTL/p1653350298061389) and this [issue on dbt](https://github.com/dbt-labs/dbt-core/issues/3580#issuecomment-929125142). 
Basically it is saying that if you have some dbt utils in a generic test inside a where clause it can break the compilation/parsing. 
Some test like below could give this error:

```

models:
  - name: shopify_order
    columns:
      - name: customer_id
        tests:
          - not_null
          - relationships:
              to: ref(any_model_with_this_column)  # test argument
              field: customer_id  # test argument
              config: # test config, must be rendered at parse time, macros other than var + env_var are not supported today
                where: "date_created < {{ dbt_utils.dateadd('day', -1, 'current_date') }}"
``` 

The issue is that the dbt_utils in the where (`where: "date_created < {{ dbt_utils.dateadd('day', -1, 'current_date') }}"`) is not interpreted correctly. 

This is the only idea I can have, so maybe just check if you have such a case of dbt_utils in a test where clause (ciuld be any dbt_utils function) and try remove it (or using basic snowflake functions - `dbt_utils.dateadd` = [DATEADD](https://docs.snowflake.com/en/sql-reference/functions/dateadd.html) etc) 


I think that might be the case and I'll need to surrender if it is not! 

PS: There is quite a lot of discussions in the slack post and in the Issue + PR on dbt :) 

Hope that's it! Have a great day

**Comment by alanmcruickshank** [comment]
Given the comment above, and that this is stale - I'm going to assume that this is resolved. Feel free to re-open the issue if it's still present in current SQLFluff versions.
